### ISSUE TITLE
[dart2js] JS-interop type tests give different results at compile time vs. runtime

### ISSUE THREAD

**Issue by fishythefish** [dart-lang/sdk#46456]
The following code appears inconsistent:

```dart
import 'package:js/js.dart';

@JS()
class A {}

@JS()
class B {}

A fa() => throw 'A';

void main() {
  print(A() is B); // #1: true
  print(fa is B Function()); // #2: false
  
  //B();
  print(test<B>(A())); // #3: false?
  print(test<B Function()>(fa)); // #4: false?
}

@pragma('dart2js:noInline')
bool test<T>(dynamic x) => x is T;
```

Checks 1 and 2 print `true` and `false` respectively, while 3 and 4 both print `false`. However, if the commented line is uncommented, then 3 and 4 both print `true`.

Firstly, why does uncommenting that line cause 3 and 4 to change? Well, without that line, `B` isn't live, so we assume that `x is B` or equivalent will always be `false`. Therefore, when we generate the subtyping metadata, we omit `B` as a supertype of `JavaScriptObject` to save space. Uncommenting the line makes `B` live, so `B` appears as a supertype of `JavaScriptObject`, and suddenly `A` starts behaving like a subtype of `B`. Note that other code like `eval` would have the same effect.

Secondly, why do checks 1 and 2 differ from 3 and 4? This is because `test` causes the latter checks to be performed at runtime, while the direct `is` checks are evaluated during SSA. (Note that not all direct `is` checks are evaluated during SSA - a counterexample is `<A>[] is List<B>`.)

Thirdly, why do checks 1 and 2 differ from each other? I'm not actually sure; I need to step through some of the SSA/abstract value domain code to see how we're simplifying these checks. It's possible that one hits a code path that recognizes interop types and the other doesn't. Maybe there's a more general bug lurking somewhere.

Finally, what would it take to reconcile these with each other (and presumably produce `true` in all cases)? It's probably feasible to get 2 to agree with 1; it just takes some debugging to figure out why they differ. Getting 3 and 4 to have stable values is trickier since we don't want to emit more subtyping metadata than we have to. There might be a middle ground: in this case, rather than requiring `B` to be live, it should be enough that it's an interop type which appears in a covariant position on the RHS of a type test. This may cause a code size increase, so we'll have to see how much that is for our customers.

/cc @sigmundch @srujzs

**Comment by sigmundch** [comment]
Note, the code above is also giving different results due to tree-shaking, if you add:

```dart
@JS()
external get markAllInteropAsLive;

main() {
   print(markAllInteropAsLive);
   ...
}
```

Then the runtime checks change behavior. That is, (1) and (2) continue to print `true` and `false`, but (3) and (4) both print `true`.

**Comment by sigmundch** [comment]
My apologies, I wrote my comment without fully reading your original message :facepalm:

**Comment by srujzs** [comment]
> print(A() is B); // #1: true
> print(fa is B Function()); // #2: false

This is interesting that the second fails, I'm assuming a similar check fails in a contravariant position as well? We have some checks in https://github.com/dart-lang/sdk/blob/master/tests/lib/js/extends_test/extends_test_util.dart#L198 that test subtyping with inheritance on `package:js` types, but these reflect the subtyping behavior you would expect if you treated the `package:js` classes like Dart classes and not the behavior you're looking for.

> There might be a middle ground: in this case, rather than requiring B to be live, it should be enough that it's an interop type which appears in a covariant position on the RHS of a type test.

Liveness has kind of been a pain point when it comes to interop types or `@Native` types. Kind of unrelated, I remember discussing something similar here where we could say that if a `@Native` type is referred to, even if it isn't instantiated, it can be considered live. I don't remember the specifics, but I do remember when discussing this that it would be a complicated change that (like you mentioned) might bloat code size.

**Comment by fishythefish** [comment]
> This is interesting that the second fails, I'm assuming a similar check fails in a contravariant position as well?

Right, if I declare `void gb(B b) {}`, then `gb is void Function(A)` is simplified at SSA to `false` while `test<void Function(A)>(gb)` is `true` iff `B` is live.
